proposal.go

```diff
@@ -0,0 +1,219 @@
+package noble
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"slices"
+
+	"connectrpc.com/connect"
+	"cosmossdk.io/log"
+	dollarkeeper "dollar.noble.xyz/keeper"
+	dollarportaltypes "dollar.noble.xyz/types/portal"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/mempool"
+	wormholekeeper "github.com/noble-assets/wormhole/keeper"
+	wormholetypes "github.com/noble-assets/wormhole/types"
+	vaautils "github.com/wormhole-foundation/wormhole/sdk/vaa"
+
+	jester "jester.noble.xyz/api"
+)
+
+const (
+	// index to inject jester response into block
+	injectIndex = 0
+)
+
+type ProposalHandler struct {
+	logger         log.Logger
+	bApp           *baseapp.BaseApp
+	jesterCli      jester.QueryServiceClient
+	dollarKeeper   *dollarkeeper.Keeper
+	wormholeKeeper *wormholekeeper.Keeper
+
+	defaultPrepareProposalHandler sdk.PrepareProposalHandler
+	defaultProcessProposalHandler sdk.ProcessProposalHandler
+}
+
+func NewProposalHandler(
+	logger log.Logger,
+	bApp *baseapp.BaseApp,
+	mp mempool.Mempool,
+	jesterCli jester.QueryServiceClient,
+	dollarKeeper *dollarkeeper.Keeper,
+	wormholeKeeper *wormholekeeper.Keeper,
+) *ProposalHandler {
+	defaultHandler := baseapp.NewDefaultProposalHandler(mp, bApp)
+	return &ProposalHandler{
+		logger:         logger,
+		bApp:           bApp,
+		jesterCli:      jesterCli,
+		dollarKeeper:   dollarKeeper,
+		wormholeKeeper: wormholeKeeper,
+
+		defaultPrepareProposalHandler: defaultHandler.PrepareProposalHandler(),
+		defaultProcessProposalHandler: defaultHandler.ProcessProposalHandler(),
+	}
+}
+
+// PrepareProposal is called only by the proposing validator and prepares a proposal
+// for the next block.
+// It calls Jester to check if there are any outstanding VAAs. These VAAs are injected
+// as bytes into the first transaction of the block and will later be handled by the PreBlocker.
+func (h *ProposalHandler) PrepareProposal() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		log := ctx.Logger()
+
+		// Call default handler first for basic validation
+		res, err := h.defaultPrepareProposalHandler(ctx, req)
+		if err != nil {
+			return nil, fmt.Errorf("failed in default PrepareProposal handler: %w", err)
+		}
+
+		// Query Jester for VAA's
+		request := connect.NewRequest(&jester.GetVoteExtensionRequest{})
+		jRes, err := h.jesterCli.GetVoteExtension(context.Background(), request)
```

Comment:
We should consider using the sdk's context here, OR a `context.WithDeadline`.

Reply:
See: 9489deb3b6421e1718d02e630baea8af82a81149

Went with a timeout of 500 milliseconds. 
This should keep our block time unaffected while still allowing expensive queries. 

Might be too safe, open to making this shorter.